### PR TITLE
refactor the packet trait

### PR DESCRIPTION
**Pull request by drunkirishcoder**
## Description

Refactor the trait to expose only the functions that are necessary. One of the main change is split the trait into `Packet` for consuming code and `PacketBase` for protocol implementors. Also all packet header types are implementation details and now private to the crate.

## Type of change

- [x] Breaking change

### CODE REVIEW

File: core/src/packets/udp.rs [hunk 1]
```diff
@@ -290,41 +315,38 @@ impl<E: IpPacket> Packet for Udp<E> {
...
-        self.mbuf_mut().shrink(offset, len)?;
-        Ok(self.envelope)
+    fn deparse(self) -> Self::Envelope {
+        self.envelope
     }
 
+    /// Reconciles the derivable attributes against the changes made to the
+    /// packet.
+    ///
+    /// * [`length`] is set to the total length of the header and the payload.
```
Comment (zeeshanlakhani) [inline core/src/packets/udp.rs:325]: ❤️ these comments for each reconcile fn.

File: core/src/packets/icmp/v4/mod.rs [hunk 2]
```diff
@@ -294,10 +303,16 @@ impl Icmpv4Payload for () {
...
 ///
 /// When using the associated derive macro, the payload struct implementation
-/// must provide an private implementation of the `cascade` function.
+/// must provide an private implementation of the `reconcile` function.
 ///
 /// [`Packet`]: crate::packets::Packet
-pub trait Icmpv4Packet<P: Icmpv4Payload>: Packet<Header = Icmpv4Header, Envelope = Ipv4> {
+pub trait Icmpv4Packet<P: Icmpv4Payload>: Packet<Envelope = Ipv4> {
+    /// Returns a reference to the header.
+    fn header(&self) -> &Icmpv4Header;
```
Comment (zeeshanlakhani) [inline core/src/packets/icmp/v4/mod.rs:311]: just for comment, as we discussed this, but these "public" trait fns for header will eventually become privately encapsulated, right @drunkirishcoder? (in another PR)

File: core/src/packets/icmp/v4/mod.rs [hunk 3]
```diff
@@ -123,35 +124,44 @@ impl Packet for Icmpv4<()> {
...
+            payload: self.payload,
+            offset: self.offset,
+        }
     }
 
-    #[doc(hidden)]
+    /// Parses the envelope's payload as an `ICMPv4` packet.
+    ///
+    /// [`Ipv4::protocol`] must be set to [`ProtocolNumbers::Icmpv4`].
+    /// Otherwise, returns a parsing error.
```
[the remark this answers is not in the document]
Comment (drunkirishcoder) [reply core/src/packets/icmp/v4/mod.rs:150]: the error is not that interesting, so I'm thinking no.

File: core/src/packets/icmp/v4/mod.rs (hunk 2 above)
Comment (drunkirishcoder) [reply core/src/packets/icmp/v4/mod.rs:311]: yeah trying to make icmp headers internal to the crate as well.

File: core/src/packets/mod.rs [hunk 4]
```diff
@@ -172,49 +177,94 @@ pub trait Packet: PacketBase {
...
+    /// [`Mbuf`]: Mbuf
+    #[inline]
     fn reset(self) -> Mbuf
     where
         Self: Sized,
     {
         self.deparse().reset()
     }
+
+    /// Reconciles the derivable attributes against the changes made to the
```
[the remark this answers is not in the document]
Comment (drunkirishcoder) [reply core/src/packets/mod.rs:248]: change everything to "derivable header fields".

**Pull request merged**